File: codes/dart/006_arrays.dart
import "dart:convert";

void main() {
    dynamic prettyArrayOfPrimitives(dynamic anything) {
        final jsonEncoder = new JsonEncoder();
        return jsonEncoder.convert(anything).split(',').join(', ');
    }

    dynamic prettyJsonStringify(dynamic anything) {
        final jsonEncoder = new JsonEncoder.withIndent("    ");
        return jsonEncoder.convert(anything);
    }

    // in Dart, JavaScript-like Array is called List

    final fruits = <dynamic>["apple", "mango", "orange"];
    print("fruits: ${prettyArrayOfPrimitives(fruits)}");

    print("fruits, length: ${fruits.length}");
    // fruits, length: 3

    print("fruits, get mango: ${fruits[1]}");
    // fruits, get mango: mango

    print("fruits, first element: ${fruits[0]}");
    // fruits, first element: apple

    print("fruits, first element: ${fruits.first}");
    // fruits, first element: apple

    print("fruits, last element: ${fruits[fruits.length - 1]}");
    // fruits, last element: orange

    print("fruits, last element: ${fruits.last}");
    // fruits, last element: orange

    fruits.asMap().forEach((dynamic arrayItemIndex, dynamic arrayItem) => print("fruits, forEach loop, index: ${arrayItemIndex}, value: ${arrayItem}"));
    // fruits, forEach loop, index: 0, value: apple
    // fruits, forEach loop, index: 1, value: mango
    // fruits, forEach loop, index: 2, value: orange

    for (final objectEntry in fruits.asMap().entries) {
        final arrayItemIndex = objectEntry.key;
        final arrayItem = objectEntry.value;
        print("fruits, for loop, index: ${arrayItemIndex}, value: ${arrayItem}");
    }
    // fruits, for loop, index: 0, value: apple
    // fruits, for loop, index: 1, value: mango
    // fruits, for loop, index: 2, value: orange

    // in Dart, JavaScript-like Array of Objects is called List of Maps

    final products = <dynamic>[
        <String, dynamic>{
            "id": "P1",
            "name": "bubble gum"
        },
        <String, dynamic>{
            "id": "P2",
            "name": "potato chips"
        }
    ];
    print("products: ${prettyJsonStringify(products)}");

    products.asMap().forEach((dynamic arrayItemIndex, dynamic arrayItem) {
        arrayItem.entries.toList().asMap().forEach((dynamic objectEntryIndex, dynamic objectEntry) {
            final objectKey = objectEntry.key;
            final objectValue = objectEntry.value;
            print("products, forEach loop, array item index: ${arrayItemIndex}, object iteration/entry index: ${objectEntryIndex}, key: ${objectKey}, value: ${objectValue}");
        });
    });
    // products, forEach loop, array item index: 0, object iteration/entry index: 0, key: id, value: P1
    // products, forEach loop, array item index: 0, object iteration/entry index: 1, key: name, value: bubble gum
    // products, forEach loop, array item index: 1, object iteration/entry index: 0, key: id, value: P2
    // products, forEach loop, array item index: 1, object iteration/entry index: 1, key: name, value: potato chips

    for (final objectEntry in products.asMap().entries) {
        final arrayItemIndex = objectEntry.key;
        final arrayItem = objectEntry.value;
        arrayItem.entries.toList().asMap().forEach((dynamic objectEntryIndex, dynamic objectEntry) {
            final objectKey = objectEntry.key;
            final objectValue = objectEntry.value;
            print("products, for loop, array item index: ${arrayItemIndex}, object iteration/entry index: ${objectEntryIndex}, key: ${objectKey}, value: ${objectValue}");
        });
    }
    // products, for loop, array item index: 0, object iteration/entry index: 0, key: id, value: P1
    // products, for loop, array item index: 0, object iteration/entry index: 1, key: name, value: bubble gum
    // products, for loop, array item index: 1, object iteration/entry index: 0, key: id, value: P2
    // products, for loop, array item index: 1, object iteration/entry index: 1, key: name, value: potato chips

    products.asMap().forEach((dynamic arrayItemIndex, dynamic arrayItem) {
        for (final objectEntry in arrayItem.entries.toList().asMap().entries) {
            final objectEntryIndex = objectEntry.key;
            final objectKey = objectEntry.value.key;
            final objectValue = objectEntry.value.value;

            print("products, forEach loop, array item index: ${arrayItemIndex}, object iteration/entry index: ${objectEntryIndex}, key: ${objectKey}, value: ${objectValue}");
        }
    });
    // products, forEach loop, array item index: 0, object iteration/entry index: 0, key: id, value: P1
    // products, forEach loop, array item index: 0, object iteration/entry index: 1, key: name, value: bubble gum
    // products, forEach loop, array item index: 1, object iteration/entry index: 0, key: id, value: P2
    // products, forEach loop, array item index: 1, object iteration/entry index: 1, key: name, value: potato chips

    for (final objectEntry in products.asMap().entries) {
        final arrayItemIndex = objectEntry.key;
        final arrayItem = objectEntry.value;
        for (final objectEntry in arrayItem.entries.toList().asMap().entries) {
            final objectEntryIndex = objectEntry.key;
            final objectKey = objectEntry.value.key;
            final objectValue = objectEntry.value.value;

            print("products, for loop, array item index: ${arrayItemIndex}, object iteration/entry index: ${objectEntryIndex}, key: ${objectKey}, value: ${objectValue}");
        }
    }
    // products, for loop, array item index: 0, object iteration/entry index: 0, key: id, value: P1
    // products, for loop, array item index: 0, object iteration/entry index: 1, key: name, value: bubble gum
    // products, for loop, array item index: 1, object iteration/entry index: 0, key: id, value: P2
    // products, for loop, array item index: 1, object iteration/entry index: 1, key: name, value: potato chips
}

File: codes/csharp/006_arrays.cs
using System; // Console, Func<>, Action<>, Action
using System.Linq; // Enumerable
using System.Collections.Generic; // Dictionary<>, List<>, KeyValuePair<>, IEnumerable<>

class Program {
    static void Main(string[] Args) {
        dynamic IsNumeric = (Func<dynamic, bool>)((dynamic Anything) => (Anything is sbyte || Anything is byte || Anything is short || Anything is ushort || Anything is int || Anything is uint || Anything is long || Anything is ulong || Anything is float || Anything is double || Anything is decimal));

        string JsonStringify(dynamic Anything, bool Pretty = false, string Indent = "    ") {
            int IndentLevel = 0;
            string JsonStringifyInner(dynamic AnythingInner, string IndentInner) {
                if (AnythingInner == null) return "null";
                if (AnythingInner is string) return "\"" + (string)AnythingInner + "\"";
                if ((IsNumeric(AnythingInner) == true) || AnythingInner is bool) return AnythingInner.ToString().Replace(",", ".");
                if (AnythingInner is List<dynamic>) {
                    IndentLevel += 1;
                    string Result = ((Pretty == true) ? ("[" + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel))) : "[");
                    int ArrayItemIndex = 0;
                    foreach (dynamic ArrayItem in (List<dynamic>)AnythingInner) {
                        Result += JsonStringifyInner(ArrayItem, IndentInner);
                        if ((ArrayItemIndex + 1) != ((List<dynamic>)AnythingInner).Count) Result += ((Pretty == true) ? ("," + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel))) : ", ");
                        ArrayItemIndex += 1;
                    }
                    IndentLevel -= 1;
                    Result += ((Pretty == true) ? (Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) + "]") : "]");
                    return Result;
                }
                if (AnythingInner is Dictionary<string, dynamic>) {
                    IndentLevel += 1;
                    string Result = ((Pretty == true) ? ("{" + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel))) : "{");
                    int ObjectIterationIndex = 0;
                    foreach (KeyValuePair<string, dynamic> ObjectEntry in (Dictionary<string, dynamic>)AnythingInner) {
                        string ObjectKey = ObjectEntry.Key;
                        dynamic ObjectValue = ObjectEntry.Value;
                        Result += "\"" + ObjectKey + "\": " + JsonStringifyInner(ObjectValue, IndentInner);
                        if ((ObjectIterationIndex + 1) != ((Dictionary<string, dynamic>)AnythingInner).Count) Result += ((Pretty == true) ? ("," + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel))) : ", ");
                        ObjectIterationIndex += 1;
                    }
                    IndentLevel -= 1;
                    Result += ((Pretty == true) ? (Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) + "}") : "}");
                    return Result;
                }
                return "null";
            };
            return JsonStringifyInner(Anything, Indent);
        }

        int ArrayItemIndex = 0;

        // in C#, JavaScript-like Array is called List

        dynamic Fruits = new List<dynamic>() {"apple", "mango", "orange"};
        Console.WriteLine($"Fruits: {JsonStringify(Fruits)}");

        Console.WriteLine($"Fruits, length: {Fruits.Count}");
        // Fruits, length: 3

        Console.WriteLine($"Fruits, get mango: {Fruits[1]}");
        // Fruits, get mango: mango

        Console.WriteLine($"Fruits, first element: {Fruits[0]}");
        // Fruits, first element: apple

        Console.WriteLine($"Fruits, last element: {Fruits[Fruits.Count - 1]}");
        // Fruits, last element: orange

        foreach (dynamic ArrayItem in (List<dynamic>)Fruits) {
            Console.WriteLine($"Fruits, forEach loop, index: {ArrayItemIndex}, value: {ArrayItem}");
            ArrayItemIndex += 1;
        }
        // Fruits, forEach loop, index: 0, value: apple
        // Fruits, forEach loop, index: 1, value: mango
        // Fruits, forEach loop, index: 2, value: orange

        // in C#, JavaScript-like Array of Objects is called List of Dictionaries

        dynamic Products = new List<dynamic>() {
            new Dictionary<string, dynamic>() {
                {"id", "P1"},
                {"name", "bubble gum"}
            },
            new Dictionary<string, dynamic>() {
                {"id", "P2"},
                {"name", "potato chips"}
            }
        };
        Console.WriteLine($"Products: {JsonStringify(Products, Pretty: true)}");

        foreach (dynamic ArrayItem in (List<dynamic>)Products) {
            int ObjectIterationIndex = 0;
            foreach (KeyValuePair<string, dynamic> ObjectEntry in (Dictionary<string, dynamic>)ArrayItem) {
                string ObjectKey = ObjectEntry.Key;
                dynamic ObjectValue = ObjectEntry.Value;
                Console.WriteLine($"Products, forEach loop, array item index: {ArrayItemIndex}, object iteration/entry index: {ObjectIterationIndex}, key: {ObjectKey}, value: {ObjectValue}");
                ObjectIterationIndex += 1;
            }
            ArrayItemIndex += 1;
        }
        // Products, forEach loop, array item index: 0, object iteration/entry index: 0, key: id, value: P1
        // Products, forEach loop, array item index: 0, object iteration/entry index: 1, key: name, value: bubble gum
        // Products, forEach loop, array item index: 1, object iteration/entry index: 0, key: id, value: P2
        // Products, forEach loop, array item index: 1, object iteration/entry index: 1, key: name, value: potato chips
    }
}
